llarp/router/rc_lookup_handler.cpp

```diff
@@ -0,0 +1,323 @@
+    whitelistRouters.erase(router);
+  }
+
+  void
+  RCLookupHandler::SetRouterWhitelist(const std::vector< RouterID > &routers)
+  {
+    util::Lock l(&_mutex);
+
+    whitelistRouters.clear();
+    for (auto &router : routers)
+    {
+      whitelistRouters.emplace(router);
+    }
+
+    LogInfo("lokinet service node list now has ", whitelistRouters.size(),
+            " routers");
+  }
+
+  void
+  RCLookupHandler::GetRC(const RouterID &router, RCRequestCallback callback)
+  {
+    RouterContact remoteRC;
+
+    if(_nodedb->Get(router, remoteRC))
+    {
+      if (callback)
+      {
+        callback(router, &remoteRC, RCRequestResult::Success);
+      }
+      FinalizeRequest(router, &remoteRC, RCRequestResult::Success);
+      return;
+    }
+
+    bool shouldDoLookup = false;
+
+    {
+      util::Lock l(&_mutex);
+
+      auto itr_pair = pendingCallbacks.emplace(router, CallbacksQueue{});
+
+      if (callback)
+      {
+        itr_pair.first->second.push(callback);
+      }
+      shouldDoLookup = itr_pair.second;
+    }
+
+    if (shouldDoLookup)
+    {
+
+      auto fn = std::bind(&RCLookupHandler::HandleDHTLookupResult,
+                            this,
+                            router,
+                            std::placeholders::_1);
+
+      // if we are a client try using the hidden service endpoints
+      if(!isServiceNode)
+      {
+        bool sent = false;
+        LogInfo("Lookup ", router, " anonymously");
+        _hiddenServiceContext->ForEachService(
+            [&](const std::string &,
+                const std::shared_ptr< service::Endpoint > &ep) -> bool {
+              const bool success = ep->LookupRouterAnon(router, fn);
+              sent               = sent || success;
+              return !success;
+            });
+        if(sent)
+          return;
+        LogWarn("cannot lookup ", router, " anonymously");
+      }
+
+      if (!_dht->impl->LookupRouter(router, fn))
+      {
+        FinalizeRequest(router, nullptr, RCRequestResult::RouterNotFound);
+      }
+    }
+  }
+
+  bool
+  RCLookupHandler::RemoteIsAllowed(const RouterID &remote) const
+  {
+    if(_strictConnectPubkeys.size() &&
+       _strictConnectPubkeys.count(remote) == 0 &&
+       !RemoteInBootstrap(remote))
+    {
+      return false;
+    }
+
+    util::Lock l(&_mutex);
+
+    if(useWhitelist
+       && whitelistRouters.find(remote) == whitelistRouters.end())
+    {
+      return false;
+    }
+
+    return true;
+  }
+
+  bool
+  RCLookupHandler::CheckRC(const RouterContact &rc) const
+  {
+    if (not RemoteIsAllowed(rc.pubkey))
+    {
+      return false;
+    }
+
+    if (not rc.Verify(_dht->impl->Now()))
+    {
+      return false;
+    }
+
+    // update dht if required
+    if(_dht->impl->Nodes()->HasNode(dht::Key_t{rc.pubkey}))
+    {
+      _dht->impl->Nodes()->PutNode(rc);
+    }
+
+    // update nodedb if required
+    if (rc.IsPublicRouter())
+    {
+      _nodedb->UpdateAsyncIfNewer(rc);
```

Comment:
this doesn't update the inserted timestamp so it either should be `InsertAsync` or make `UpdateAsyncIfNewer` change insertion timestamps.

Reply:
I think changing UpdateAsyncIfNewer is cleaner, which do you prefer?

Reply:
changing `UpdateAsyncIfNewer` works fine

Reply:
Changed.